js/src/components/tree-select-control/index.js

```diff
@@ -1,56 +1,190 @@
 /**
  * External dependencies
  */
-import { useState } from '@wordpress/element';
+import { useMemo, useState } from '@wordpress/element';
 import classnames from 'classnames';
-// import { __, _n, sprintf } from '@wordpress/i18n';
-// import { Component, createElement } from '@wordpress/element';
-// import { debounce, escapeRegExp, identity, noop } from 'lodash';
-// import PropTypes from 'prop-types';
-// import { withFocusOutside, withSpokenMessages } from '@wordpress/components';
-// import { withInstanceId, compose } from '@wordpress/compose';
+// eslint-disable-next-line import/no-extraneous-dependencies,@woocommerce/dependency-group,@wordpress/no-unsafe-wp-apis
+import { __experimentalUseFocusOutside as useFocusOutside } from '@wordpress/compose';
 
 /**
  * Internal dependencies
  */
-import List from './list';
 import Control from './control';
+import List from './list';
 import './index.scss';
 
-const TreeSelectControl = ( props ) => {
-	const { className, disabled, value = [], onChange = () => {} } = props;
-	const [ isExpanded, setIsExpanded ] = useState( true );
+/**
+ * The Option type Object. This is how we send the options to the selector.
+ *
+ * @typedef { Object } Option
+ * @property {string} id The unique ID for the option
+ * @property {string} name The name for the option
```

Comment:
I'm fine with both options. Not sure tho if In see the advantage of one over the other, so I rely on your preference about that :) 

It was built like this because I used [Tree Select](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/tree-select) as a reference, as we discussed.

> Also, just in case there are multiple tree nodes having the same `value` 

Is this allowed? In my mind was unique values, otherwise, you don't really know which one you selected right? 

I think you're considering something like:

**Food selector**

Vegan
  -  Green Beans

Vegetarian 
  -  Green Beans
  -  Eggs 

Meat
  - Spanish Ham
  
In which an option like Green Beans, can be distributed in several groups, right?

Reply:
> It was built like this because I used [Tree Select](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/tree-select?rgh-link-date=2022-02-23T08%3A30%3A06Z) as a reference, as we discussed.

Haha, actually I always prefer to use Ant Design as reference ([TreeSelect](https://ant.design/components/tree-select/)), the one you mentioned is from `@wordpress/components`. 😄 But thanks for bringing that up, at least it is a reference too. 

> In which an option like Green Beans, can be distributed in several groups, right?

Yeah, and "Vegan > Green beans" and "Vegetarian > Green beans" should have unique values. 

> Also, just in case there are multiple tree nodes having the same `value` (they may have different parent tree node), we might want to consider to let the tree node have a `key` property that is unique across the whole tree. This may be too far-fetch though. Just thinking out loud.

I guess I was overthinking too much. 😅 We can stick with something simple, assume unique values across the whole tree (like you said), and go with that.

Reply:
> 💅 ❔ Instead of `id` and `name`, I'm thinking it may be better to use `value` and `label`, similar to [HTML Option element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/option). What do you think?

Sorry, I had more thinking about this and wanted to get back to this but I forgot and missed it until now. 😅 

After some more thoughts, I'm still in favor of `value` and `label` over `id` and `name`, because:

- `id` might be misleading or confusing the consumer, e.g. "will the id be in the HTML `id` attribute output like `<option id="123">Text here</option>`?" or "is it related to the instance id?" (you worked on instance id in https://github.com/woocommerce/google-listings-and-ads/pull/1243#discussion_r814565752). However, the reality is that our `id` here has nothing to do with HTML `id` attribute.
- `name` might be misleading or confusing the consumer, e.g. "is name related to HTML `name` attribute like `<input type="text" name="name1">`?". Notice that the HTML `name` attribute is not user-facing, but here we are using `name` to display text to the users. This could be a mismatch on consumer's expectation.
- On the other hand, `value` is clearer in its semantic, the consumer might reasonably expect it to behave like `<option value="value1">Text1</option>`. Even though we don't generate the HTML output exactly like that, it would still match the consumer's expectation - it is not user-facing, and it is used to indicate the options selected by the users.
- `label` is clearer in its semantic. When we say "label", we know it is user-facing, users will be seeing it. 

WDYT? 🙂